erts/emulator/beam/beam_emu.c

```diff
@@ -7035,6 +7047,66 @@ new_map(Process* p, Eterm* reg, BeamInstr* I)
 }
 
 static Eterm
+new_map_lit(Process* p, Eterm* reg, Uint* n_exp, BeamInstr* I)
+{
+    Eterm* keys = tuple_val(Arg(3));
+    Uint n = arityval(*keys);
+    Uint need = n + 1 /* hdr */ + 1 /*size*/ + 1 /* ptr */ + 1 /* arity */;
+    Uint i;
+    BeamInstr *ptr;
+    flatmap_t *mp;
+    Eterm *mhp,*thp;
+    Eterm *E;
+    ErtsHeapFactory factory;
+
+    *n_exp = n;
+    ptr = &Arg(4);
+
+    if (n > MAP_SMALL_MAP_LIMIT) {
```

Comment:
Why is this test done at run-time?

It seems to me that it would be to do this test in the loader and only let this instruction handle small maps.